File: compiler/src/main/java/edu/cornell/cs/apl/viaduct/algebra/solver/JoinEdge.java
package edu.cornell.cs.apl.viaduct.algebra.solver;

import edu.cornell.cs.apl.viaduct.algebra.Lattice;
import edu.cornell.cs.apl.viaduct.util.dataflow.DataFlowEdge;
import org.jgrapht.graph.DefaultEdge;

/** Joins the incoming value with a constant before passing it on. */
final class JoinEdge<A extends Lattice<A>> extends DefaultEdge implements DataFlowEdge<A> {
  private final A constant;

  JoinEdge(A constant) {
    this.constant = constant;
  }

  @Override
  public A propagate(A in) {
    return constant.join(in);
  }

  @Override
  public String toString() {
    return constant + " ∨ _";
  }
}

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/algebra/solver/JoinEdge.kt
package edu.cornell.cs.apl.viaduct.algebra.solver

import edu.cornell.cs.apl.viaduct.algebra.Lattice
import edu.cornell.cs.apl.viaduct.util.dataflow.DataFlowEdge
import org.jgrapht.graph.DefaultEdge

/** Joins the incoming value with a constant before passing it on. */
internal class JoinEdge<A : Lattice<A>>(private val constant: A) : DefaultEdge(), DataFlowEdge<A> {
    override fun propagate(input: A): A {
        return constant.join(input)
    }

    override fun toString(): String {
        return "$constant ∨ _"
    }
}
